File: code/ex713.java
// Sliding window

class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1)    return 0;
        int n = nums.length;
        
        int start = 0, end = 0, res = 0;
        int product = 1;
        while (end < n) {
            product *= nums[end];
            while (product >= k)    product /= nums[start++];
            res += end - start + 1;
            end++;
        }
        
        return res;
    }
}

File: code/ex713.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int n = nums.size();
        
        int lo = 0, hi = 0, pro = 1, res = 0;
        while (hi < n) {
            pro *= nums[hi];
            while (lo <= hi && pro >= k) {
                pro /= nums[lo++];
            }
            res += hi - lo + 1;
            hi++;
        }
        
        return res;
    }
};